plutus-ledger/src/Ledger/Tx/Internal.hs

```diff
@@ -70,15 +90,72 @@ instance Pretty TxIn where
                             _ -> mempty
                 in hang 2 $ vsep ["-" <+> pretty txInRef, rest]
 
+-- | A transaction input that spends a "pay to public key" output, given the witness.
+pubKeyTxIn :: TxOutRef -> TxIn
+pubKeyTxIn r = TxIn r (Just ConsumePublicKeyAddress)
+
+-- | A transaction input that spends a "pay to script" output, given witnesses.
+scriptTxIn :: LedgerPlutusVersion -> TxOutRef -> Validator -> Redeemer -> Datum -> TxIn
+scriptTxIn pv ref v r d = TxIn ref . Just $ ConsumeScriptAddress pv v r d
+
+-- | The type of a transaction input. Contains redeemer if consumes a script.
+data TxInputType =
+      TxConsumeScriptAddress !LedgerPlutusVersion !Redeemer !ValidatorHash !DatumHash -- ^ A transaction input that consumes a script address with the given validator, redeemer, and datum.
+    | TxConsumePublicKeyAddress -- ^ A transaction input that consumes a public key address.
+    | TxConsumeSimpleScriptAddress -- ^ Consume a simple script
+    deriving stock (Show, Eq, Ord, Generic)
+    deriving anyclass (ToJSON, FromJSON, Serialise, NFData)
+
+-- | A transaction input, consisting of a transaction output reference and an input type.
+-- Differs with TxIn by: TxIn *maybe* contains *full* data witnesses, TxInput always contains redeemer witness, but datum/validator hashes.
+data TxInput = TxInput {
```

Comment:
Why not just change the representation of `TxIn` instead of introducing a new type?

Reply:
`TxIn` remains used, wanted to make the change the easiest way. Imo best in the future is to remove `Maybe` in the `txInType` field of `TxIn` (like  `TxInputWitnessed`). I haven't found this field being optional as needed.